test/integration/hotrestart_handoff_test.py

```diff
@@ -0,0 +1,262 @@
+UPSTREAM_FAST_PORT = 54322
+UPSTREAM_HOST = random_loopback_host()
+ENVOY_HOST = UPSTREAM_HOST
+ENVOY_PORT = 54323
+ENVOY_ADMIN_PORT = 54324
+SOCKET_PATH = "@envoy_domain_socket"
+SOCKET_MODE = 0
+ENVOY_BINARY = "./test/integration/hotrestart_main"
+
+# This log config makes logs interleave with other test output, which
+# is useful since with all the async operations it can be hard to figure
+# out what's happening.
+log = logging.getLogger()
+log.level = logging.INFO
+_stream_handler = logging.StreamHandler(sys.stdout)
+log.addHandler(_stream_handler)
+
+
+class Upstream:
+    # This class runs a server which takes an http request to
+    # path=/ and responds with "start\n" [three second pause] "end\n".
+    # This allows us to test that during hot restart an already-opened
+    # connection will persist.
+    # If initialized with True it will instead respond with
+    # "fast instance" immediately.
+    def __init__(self, fast_version=False):
+        self.port = UPSTREAM_FAST_PORT if fast_version else UPSTREAM_SLOW_PORT
+        self.app = web.Application()
+        self.app.add_routes([
+            web.get('/', self.fast_response) if fast_version else web.get('/', self.slow_response),
+        ])
+
+    async def start(self):
+        self.runner = web.AppRunner(self.app, handle_signals=False)
+        await self.runner.setup()
+        site = web.TCPSite(self.runner, host=UPSTREAM_HOST, port=self.port)
+        await site.start()
+
+    async def stop(self):
+        await self.runner.shutdown()
+        await self.runner.cleanup()
+        log.debug("runner cleaned up")
+
+    async def fast_response(self, request):
+        return web.Response(
+            status=200, reason='OK', headers={'content-type': 'text/plain'}, body='fast instance')
+
+    async def slow_response(self, request):
+        log.debug("slow request received")
+        response = web.StreamResponse(
+            status=200, reason='OK', headers={'content-type': 'text/plain'})
+        await response.prepare(request)
+        await response.write(b"start\n")
+        await asyncio.sleep(3)
+        await response.write(b"end\n")
+        await response.write_eof()
+        return response
+
+
+async def http_request(url):
+    async with ClientSession() as session:
+        retries = 5
+        while retries:
+            try:
+                retries -= 1
+                async with session.get(url) as response:
+                    async for line in response.content:
+                        yield line
+                    return
+            except client_exceptions.ClientConnectorError as e:
+                if not retries:
+                    raise e
+                await asyncio.sleep(0.2)
+                if retries <= 3:
+                    log.debug(f"retrying request ({retries} remain)\n")
+
+
+def make_envoy_config_yaml(upstream_port, file_path):
+    with open(file_path, "w") as file:
+        file.write(
+            f"""
+admin:
```

Comment:
can we put this in a yaml file it will be more readable and then it will get linted properly - we will need to  exclude it from config checking in that case

Reply:
I'd rather not because it would have to be a template that you load and then modify with replace functions and re-save (which is what the shell version of hotrestart_test does, and it's gross and makes it much harder to follow what the test is actually doing). You'd gain yaml linting but lose automatic "the values you're substituting actually exist" checking and readability (and also half the yaml linting because it's a template not valid values).

Reply:
understood re editor